[WIP] Test framework

DigitalSmile opened the pull request:
# Pi4J testing framework for the win

## Overview

Pi4J lacks proper tests within it's usage. Not only unit in core tests (which are good to test some basic logic), but provider unit, integration and performance tests. Moreover, new FFM API relies on the kernel, that is run within Linux OS. That's essentially means we have to keep an eye on what is changing in kernel driver/syscall development. To keep all this in place, I propose the following framework.

## Unit tests

~~We need to use mock approach to test raw methods, that they works as expected (e.g. Mockito in FFM https://github.com/Pi4J/pi4j/tree/develop/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/mocks).
Several layers of unit testing approach:~~
- ~~core unit tests, located at https://github.com/Pi4J/pi4j/tree/develop/pi4j-core/src/test/java/com/pi4j~~
- ~~FFM and any other provider (e.g. https://github.com/Pi4J/pi4j/tree/develop/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit)~~

~~In this scope, we need to remove `pi4j-test` module and move corresponding unit  and integration tests to core.~~

It's better to leave unit tests here for now.

## Model tests

Model tests are needed to make sure any interop structure are not broken.
- FFM model tests (https://github.com/Pi4J/pi4j/tree/develop/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/model)

Same should be applied to any other plugins we offer.

### Todo

- [ ] Add model tests to pi4j-core (if necessary)

## Integration tests

This tests should check if interop and syscalls are not broken. Main idea is to write simple mock (or echo) drivers for each provider and test syscalls. I have created some drivers FFM provider within this PR:

- SPI interface
    - driver - https://github.com/Pi4J/pi4j/tree/test-framework/plugins/pi4j-plugin-ffm/src/test/native/spi
    - tests - https://github.com/Pi4J/pi4j/blob/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/SPITest.java
- I2C
    - driver - https://github.com/Pi4J/pi4j/tree/test-framework/plugins/pi4j-plugin-ffm/src/test/native/i2c
    - I2C with ioctl - https://github.com/Pi4J/pi4j/blob/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CDirectTest.java
    - I2C with file - https://github.com/Pi4J/pi4j/blob/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CFileTest.java
    - I2C with SMBus - https://github.com/Pi4J/pi4j/blob/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CSMBusTest.java
- PWM
    - driver - https://github.com/Pi4J/pi4j/tree/test-framework/plugins/pi4j-plugin-ffm/src/test/native/pwm
    - tests - https://github.com/Pi4J/pi4j/blob/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/PwmTest.java
- GPIO **TBD**
- ~~Serial **TBD**~~

This drivers requires a simple build with `make` and current loaded kernel headers (I have provided comments with source files, though there are really very simple). We can think of expanding this tests to all recent stable linux kernels in future.
Each driver is built and loaded with the test in java and removed upon test completion. This works in linux machines only, so a special annotation is used not to run them on other systems.

### Todo

- [x] Create SPI driver and fix tests
- [x] Create PWM driver and fix tests
- [x] Create I2C driver and fix tests
- [x] Create GPIO driver and fix tests
- [ ] ~~Create Serial driver and fix tests~~

## Performance tests

Since we are planning to expand the support from RaspberryPi only to every linux driven SBC, It is very possible, that pi4j will be used in high concurrent and/or highload environment. So we have to control and measure periodically how our changes affects the performance of interop and overall software responsiveness. I suggest to use the same approach as in integration testing - use simple echo/mock drivers to measure how fast the communication go. JMH tests for FFM provider are located here -> https://github.com/Pi4J/pi4j/tree/test-framework/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh
They include:

- GPIO performance tests
- PWM performance tests
- I2C performance tests
- SPI performance tests
- ~~Serial performance tests~~

We have to think on comparing ourselves to other / similar libraries.

### Todo

- [ ] Create FFM API tests for GPIO (regression and performance degradation)
- [ ] Create FFM API tests for PWM (regression and performance degradation)
- [x] Create FFM API tests for SPI (regression and performance degradation)
- [x] Create FFM API tests for I2C (regression and performance degradation)
- [ ] Create FFM API tests for Serial (regression and performance degradation)
- [ ] Create comparing performance tests
- [ ] Run tests periodically and publish results (https://github.com/kitlangton/jmh-benchmark-action)

DigitalSmile committed:
Created new kernel drivers for I2C, PWM and SPI, fix tests

DigitalSmile committed:
Created new kernel drivers for I2C, PWM and SPI, fix tests

DigitalSmile committed:
Changed mock method to 'setup', write kernel drivers, fix tests, new CI config

DigitalSmile committed:
Changed mock method to 'setup', write kernel drivers, fix tests, new CI config

DigitalSmile committed:
#540 Fix DIRECT implementation of I2C write

DigitalSmile committed:
#535 Set default to I2CDirect for provider, added logging, fix SegmentAllocator in I2CMessage and RDWRData

DigitalSmile committed:
#535 Changed log level for I2C functions

FDelporte commented on .github/workflows/ci-test.yml:
```diff
@@ -0,0 +1,34 @@
+      - name: Install JDK25
+        uses: sfesenko/setup-sdkman@v1
+        with:
+          deps: java:25-amzn
```
25-zulu please :-)

FDelporte commented on .github/workflows/ci-test.yml:
```diff
@@ -16,7 +16,7 @@ jobs:
         uses: sfesenko/setup-sdkman@v1
         with:
-          deps: java:25-amzn
+          deps: java:25-zulu
```
jiepie :-)

DigitalSmile committed:
#535 Add debug logging 'if'

DigitalSmile committed:
Fix tests and driver after merge

DigitalSmile committed:
Fix tests and driver after merge

eitch commented on .github/workflows/ci-test.yml:
```diff
@@ -0,0 +1,36 @@
+name: ci-test.yml
+on:
+  push:
+    branches:
+      - "test-framework"
+
+jobs:
+  Build:
+    name: Build Snapshot
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Install dependencies
+        # run: sudo apt-get install --yes build-essential linux-headers-$(uname -r) sudo libi2c-dev gcc-arm-linux-gnueabihf  gcc-aarch64-linux-gnu libtool tree
+        run: |
+          sudo apt-get remove --purge man-db
+          sudo apt-get install --yes build-essential linux-headers-$(uname -r) sudo libi2c-dev
+      - name: Install JDK25
+        uses: sfesenko/setup-sdkman@v1
+        with:
+          deps: java:25-zulu
+      - name: Create and add current user to suitable groups
+        run: |
+          sudo addgroup spi
+          sudo adduser $USER spi
+          sudo adduser $USER i2c
+          sudo adduser $USER dialout
+          sudo -u $USER bash -c 'groups'
+      - name: Add proper udev rules and trigger udevadm
+        run: |
+          sudo cp udev.rules /etc/udev/rules.d/99-pi4j-io.rules
+          sudo udevadm control --reload-rules && sudo udevadm trigger
+      - name: Build entire Pi4J Project
+        run: |
+          sudo -u $USER bash -c 'source ~/.sdkman/bin/sdkman-init.sh && ./mvnw clean install -Pnative,cross-compile --batch-mode'
```
What does Github want here?

eitch commented on plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/impl/I2CFile.java:
```diff
@@ -82,22 +82,28 @@ public int readRegister(int register, byte[] buffer, int offset, int length) {
     @Override
     public int writeRegister(int register, byte b) {
-        return i2CBus.execute(this, (i2cFileDescriptor) -> FILE.write(i2cFileDescriptor, new byte[]{(byte) register, b}));
+        return i2CBus.execute(this, (i2cFileDescriptor) -> FILE.write(i2cFileDescriptor, new byte[]{(byte) register, b})) - 1;
```
put the -1 in a new statement, so it does not get ignored.

eitch commented on plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/impl/I2CFile.java:
```diff
@@ -82,22 +82,28 @@ public int readRegister(int register, byte[] buffer, int offset, int length) {
         System.arraycopy(register, 0, buffer, 0, register.length);
         System.arraycopy(data, 0, buffer, register.length, data.length);
-        return i2CBus.execute(this, (i2cFileDescriptor) -> FILE.write(i2cFileDescriptor, buffer));
+        return i2CBus.execute(this, (i2cFileDescriptor) -> FILE.write(i2cFileDescriptor, buffer)) - register.length;
```
dito

eitch commented on plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java:
In these checks, isn't it counter intuitive, that the user requested SMBUS, but we don't use it if it isn't supported? Shouldn't we throw an exception?

Thus if the user requests a specific impl, then we should use that and throw exceptions if something isn't as expected.

eitch commented on plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/spi/SpiFFM.java:
```diff
@@ -127,4 +133,21 @@ private void checkClosed() {
         }
     }
+
+    public static void main(String[] args) throws IOException, InterruptedException {
```
Why do we have a main class in production code? This should probably go to a test, no?

eitch commented on plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/GPIOPerformanceTest.java:
```diff
@@ -91,85 +87,85 @@ public void testFFMOutputRoundTrip() {
-        pin.state();
-        pi4j.shutdown();
-    }
+//    @Benchmark
```
We shouldn't have commented out code. Either use a deactivation mechanism, or remove the code.

eitch commented on plugins/pi4j-plugin-ffm/src/test/native/i2c/build.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
```
I was hoping we could remove building with make from pi4j down the line... is there no better solution?

DigitalSmile replied:
I have answered you somewhere else - this is needed for testing only, there will be no runtime dependencies and cross-compile builds. It is running during test phase, only in linux with kernel headers as a single dependency. Very primitive and as simple as possible.

Unfortunately there is no "mock driver pack" for each interface, most of existing mock drivers are not the default part of kernel (and in CI we have prebuild custom azure kernel). But I still think it is important to have integrity kernel tests to check changes we make.

eitch commented on plugins/pi4j-plugin-ffm/pom.xml:
```diff
@@ -41,7 +41,7 @@
             <scope>test</scope>
         </dependency>
-        <dependency>
+<!--        <dependency>
```
let's  just remove the dependency block all together

eitch commented on plugins/pi4j-plugin/pom.xml:
```diff
@@ -29,11 +29,11 @@
     <!-- PROJECT MODULES <Pi4J PLUGINS> -->
     <modules>
-        <module>../pi4j-plugin-mock</module>
+<!--        <module>../pi4j-plugin-mock</module>
```
I think all of these should be activated again?

eitch reviewed (changes requested):
There are a few points

DigitalSmile commented:
@eitch this PR is work in progress, I am cleaning up the code on my way with todo list.
I am asking to review the approach, stated in first comment.

FDelporte committed:
Merge branch 'develop' into test-framework

# Conflicts:
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/pwm/PwmFFMHardware.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/SPITest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit/PWMTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/GPIOPerformanceTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/SPIPerformanceTest.java

stefanhaustein committed:
2 params, nanoseconds

stefanhaustein committed:
Support the delay between write and read only

FDelporte committed:
Merge branch 'develop' into extendingSmokeTests

taartspi committed:
update output test to use listener for pin change.  Wait one second and see if the listener saw a change. If yes, validate the pin state

taartspi committed:
Merge branch 'extendingSmokeTests' of https://github.com/Pi4J/pi4j into extendingSmokeTests

stefanhaustein committed:
remove spi mentions where they are too specific for the shared interface

stefanhaustein committed:
Merge pull request #577 from stefanhaustein/develop

Shared writeThenRead interface

FDelporte committed:
Merge pull request #574 from Pi4J/extendingSmokeTests

[WIP] Extending tests for clear steps and failed results

FDelporte committed:
Remove serial implementations and related code

FDelporte committed:
Remove serial implementations and related code

FDelporte committed:
Merge pull request #580 from Pi4J/feature/removeSerial

Remove serial implementations and related code

DigitalSmile committed:
#552 Implement writeThenRead

stefanhaustein committed:
Fix some methods inadvertedly changed from byte... to byte[] when introducing SerialCircuitIO

FDelporte committed:
Merge pull request #581 from Pi4J/552-ffm-spi-writethenread-not-implemented

#552 Implement writeThenRead

FDelporte committed:
Merge pull request #583 from stefanhaustein/develop

Convert write(byte[]) back to writer(byte...)

FDelporte committed:
Remove serial: further cleanup

FDelporte committed:
User permissions: add link to Pi4J OS repo for script

FDelporte committed:
User permissions: add link to Pi4J OS repo for script

FDelporte committed:
User permissions: add link to Pi4J OS repo for script

FDelporte committed:
User permissions: add link to Pi4J OS repo for script

FDelporte committed:
Merge branch 'develop' into 563-ffm-pwm-channel-other-than-0-fail-to-export-after-reboot

# Conflicts:
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/pwm/FFMPwmHardware.java

FDelporte committed:
Solve merge conflicts

FDelporte committed:
Merge pull request #569 from Pi4J/563-ffm-pwm-channel-other-than-0-fail-to-export-after-reboot

#563 Added alternative implementation with WatcherService wrapper

DigitalSmile committed:
Changed deprecated DeferredDelay to Delay

FDelporte committed:
Merge pull request #584 from Pi4J/pwm-deffered-delay

Changed deprecated DeferredDelay to Delay

DigitalSmile committed:
#552 Added read delay to tx buffer, made adjustment to nano vs micro seconds.

DigitalSmile committed:
#563 Adjusted waitForPermission for every file opening.

FDelporte committed:
Merge pull request #585 from Pi4J/552-ffm-spi-writethenread-not-implemented

#552 Added read delay to tx buffer, made adjustment to nano vs micro

FDelporte committed:
Merge pull request #586 from Pi4J/563-ffm-pwm-channel-other-than-0-fail-to-export-after-reboot

#563 Adjusted waitForPermission for every file opening.

eitch committed:
[Provider] Add `newautocontext` as default test provider and update usage instructions

eitch committed:
[Config] Set `DIRECT` as the default I2C implementation in `DefaultI2CConfig`.

FDelporte committed:
Merge pull request #587 from Pi4J/feature/smoke-test-newautocontext

[New] Add `newautocontext` as default test provider and update usage instructions

FDelporte committed:
Merge pull request #588 from Pi4J/feature/ffm-i2c-default-direct

[Minor] Set `DIRECT` as the default I2C implementation in `DefaultI2CConfig`.

FDelporte committed:
Difference between BMP280 and BME280 added

FDelporte committed:
Extra writeThenRead signature to avoid duplicate code

eitch committed:
Merge pull request #589 from Pi4J/writeThenRead-constructor

Extra writeThenRead signature to avoid duplicate code

taartspi committed:
ffm i2cDirect line 104  out of bounds. smoketest remove second PWM test with freq too high. Enable spi test on FFM

eitch committed:
Merge pull request #590 from Pi4J/smoketest_spi_testing_on_FFM

ffm i2cDirect line 104  out of bounds. smoketest remove second PWM te…

FDelporte committed:
Test log output

FDelporte committed:
Test log output

taartspi committed:
move logger init into main. This allows setting the property to take effect

FDelporte committed:
Merge pull request #591 from Pi4J/smoketest_move_logger

move logger init into main. This allows setting the property to take …

taartspi committed:
linuxfs spi. Paste error  in  math converting nano to micro

FDelporte committed:
Merge pull request #592 from Pi4J/linuxfs_spi_incorrect_math

linuxfs spi. Paste error  in  math converting nano to micro

FDelporte committed:
Test SPI in two ways, should give identical results

stefanhaustein committed:
Fix #594

FDelporte committed:
Test SPI in two ways, should give identical results

DigitalSmile committed:
Merge pull request #598 from stefanhaustein/develop

Fix #594

taartspi committed:
smoketest update,  add debounce test to assist in debug

FDelporte committed:
Merge pull request #599 from Pi4J/smoketest_debounce

 smoketest update,  add debounce test to assist in debug

FDelporte committed:
Merge pull request #593 from Pi4J/v4-testing

Final test changes heading to the release of V4...

Commit:
Full link to the images, so this README file can be used in the Pi4J website

taartspi committed:
debounce testcase has improved details of testcase outcome.  PWm test, tweaked to close window where 10 or 11 flashes are counted. SPI test. Removed the test using seperate write and reds. Without the use of a dedicated GPIO to control CS from the application, write and reads in separate steps fail. Was the reason Stefan asked for writeThenRead implementation, it uses a single ioctl to perform both steps thereby keeping CS low. The testing of both means would be good to do with a chip that will support this

taartspi committed:
debounce testcase has improved details of testcase outcome.  PWm test, tweaked to close window where 10 or 11 flashes are counted. SPI test. Removed the test using seperate write and reds. Without the use of a dedicated GPIO to control CS from the application, write and reads in separate steps fail. Was the reason Stefan asked for writeThenRead implementation, it uses a single ioctl to perform both steps thereby keeping CS low. The testing of both means would be good to do with a chip that will support this
Update to i2c testcase to use readReg and writeThenRead.  Second one
needed for regression testing FFM fix

FDelporte committed:
Updating of test wiring

FDelporte committed:
Merge branch 'develop' into smoketest_debounce_detail_enhance

FDelporte committed:
Updating of test wiring

FDelporte committed:
Updating of test wiring

taartspi committed:
Merge branch 'smoketest_debounce_detail_enhance' of https://github.com/Pi4J/pi4j into smoketest_debounce_detail_enhance

# Conflicts:
#	pi4j-test/README.md
#	pi4j-test/src/main/java/com/pi4j/test/smoketest/SpiTestCase.java

DigitalSmile committed:
#597 Fix setting debounce value

DigitalSmile committed:
#596 Fix SPI multiple buffer creation

DigitalSmile committed:
#596 Fix SPI multiple buffer creation

taartspi committed:
Merge pull request #600 from Pi4J/smoketest_debounce_detail_enhance

FDelporte committed:
Merge pull request #601 from Pi4J/597-digitalinput-debounce

#597 Fix setting debounce value

FDelporte committed:
#604: code cleanup + javadoc inherits + moved parameters to internal methods

FDelporte committed:
#604: fix unit test

FDelporte committed:
Merge branch 'develop' into 604-unused-parameters

FDelporte committed:
#596: adding logging

Commit:
Adding JavaDoc

Commit:
Adding JavaDoc

DigitalSmile committed:
#596 Fix SPI multiple buffer created empty when calling 'from' MemorySegment

FDelporte committed:
Add build time

FDelporte committed:
Add build time

FDelporte committed:
Merge branch 'develop' into 604-unused-parameters

FDelporte committed:
Ignore test trace logs

FDelporte committed:
Info package versus install

FDelporte committed:
Extra debounce smoke test

FDelporte committed:
Extra info in build.properties

FDelporte committed:
Improve debounce count test

FDelporte committed:
Extra info in build.properties

FDelporte committed:
Merge branch 'develop' into test-debounce

FDelporte committed:
Cleanup

FDelporte committed:
Use longer sleep times

FDelporte committed:
Better test titles

FDelporte committed:
Use inbetween buffers

FDelporte committed:
Merge pull request #602 from Pi4J/596-spi-writethenread

#596 Fix SPI multiple buffer creation

Commit:
Merge remote-tracking branch 'origin/develop' into 604-unused-parameters

Commit:
Fix style error

Commit:
JavaDoc

taartspi committed:
New testcase to use a logic analyzer to measure debounce

taartspi committed:
New testcase to use a logic analyzer to measure debounce

taartspi committed:
Merge branch 'test-debounce' of https://github.com/Pi4J/pi4j into test-debounce

FDelporte committed:
Merge branch 'develop' into test-debounce

FDelporte committed:
Renamed test cases for consistency

FDelporte committed:
Fix incorrect BCM used in debounce monitor

FDelporte committed:
Reorder internal parameters to match inherited APIs + avoid unneeded inbetween buffers when offset is 0

FDelporte committed:
Fix unit test

FDelporte committed:
Fix unit test

taartspi committed:
Merge branch 'test-debounce' of https://github.com/Pi4J/pi4j into test-debounce

FDelporte committed:
Add software debounce to FFMDigitalInput

FDelporte committed:
Add software debounce to FFMDigitalInput

FDelporte committed:
Add software debounce to FFMDigitalInput

FDelporte committed:
Fix byte array size

FDelporte committed:
Simplifying DigitalInputDebounceTimeTestCase

FDelporte committed:
Testing DigitalInputDebounceTimeTestCase

FDelporte committed:
Testing DigitalInputDebounceTimeTestCase

FDelporte committed:
Improving smoke tests

FDelporte committed:
Improving smoke tests

FDelporte committed:
Improving smoke tests

FDelporte committed:
Improving smoke tests

FDelporte committed:
Testing smoke tests

FDelporte committed:
Merge request remarks and improved logging

FDelporte committed:
Merge request remarks and improved logging

FDelporte committed:
Merge request remarks and improved logging

FDelporte committed:
Merge request remarks: extra unit test

FDelporte committed:
Merge pull request #606 from Pi4J/test-debounce

FFM digital input debounce improvements with extra smoke tests

FDelporte committed:
Merge branch 'develop' into 604-unused-parameters

# Conflicts:
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/impl/I2CDirect.java

FDelporte committed:
Remove unneeded wire for I2C smoketest

FDelporte committed:
Added I2CWithOffsetTestCase

FDelporte committed:
Added I2CWithOffsetTestCase

taartspi committed:
smoketest wiring. If the sensor supports both i2c and spi, the CS line controls which interface is active

taartspi committed:
Add chip datasheet url's
smoketest wiring. If the sensor supports both i2c and spi, the CS line controls which interface is active

DigitalSmile committed:
#604 Simplify internalRead methods

DigitalSmile committed:
Merge pull request #605 from Pi4J/604-unused-parameters

Fix unused offset parameter in I2CDirect

stefanhaustein committed:
Fix index/size check issue

FDelporte committed:
Merge pull request #608 from stefanhaustein/develop

Fix index/size check issue

taartspi committed:
Merge branch 'smoketest_i2c_wiring' of https://github.com/Pi4J/pi4j into smoketest_i2c_wiring

taartspi committed:
debounce timing. + or - 25ms allowed, outside that range failure

FDelporte committed:
Merge pull request #609 from Pi4J/smoketest-debounce-variance

debounce timing. + or - 25ms allowed, outside that range failure

FDelporte committed:
Merge pull request #607 from Pi4J/smoketest_i2c_wiring

smoketest wiring. If the sensor supports both i2c and spi, the CS lin…

taartspi committed:
new test for spi writeThenRead with offset.  Required a fix in lunuxfs.  In ffm, I only added the required buffer size validation so more work needed.  Also added a test spi write and read,  using separate calls.  This fails on both linuxfs and ffm, i need to use the logic analyzer to make certain this is even possible with these BMP chips.  FFM needs review of any interface that include the offset parm

taartspi committed:
Fix in spi writeThenRead.   Needs review as a better way to updae the read data might exist

 new test for spi writeThenRead with offset.  Required a fix in lunuxfs.  In ffm, I only added the required buffer size validation so more work needed.  Also added a test spi write and read,  using separate calls.  This fails on both linuxfs and ffm, i need to use the logic analyzer to make certain this is even possible with these BMP chips.  FFM needs review of any interface that include the offset parm

taartspi committed:
Merge branch 'smoketest_spi_offset' of https://github.com/Pi4J/pi4j into smoketest_spi_offset

taartspi committed:
Fix missing offset in individual write and read

Merge branch 'smoketest_spi_offset' of https://github.com/Pi4J/pi4j into smoketest_spi_offset

taartspi committed:
Merge branch 'smoketest_spi_offset' of https://github.com/Pi4J/pi4j into smoketest_spi_offset

FDelporte committed:
Inherit JavaDoc

FDelporte committed:
Swapped test values?

FDelporte committed:
Cleanup

FDelporte committed:
Test result overview

FDelporte committed:
Better result logging

FDelporte committed:
Better result logging

FDelporte committed:
MR remarks

FDelporte committed:
Merge pull request #611 from Pi4J/smoketest_spi_offset

new test for spi writeThenRead with offset. Fails to fix and FFM spi offset review/code  Required a fix in lunuxf…

FDelporte committed:
Add fzz binary to avoid corrupting Fritzing wiring files

FDelporte committed:
Fix Java version

eitch committed:
[Project] Update project version to 4.1.0-SNAPSHOT

FDelporte committed:
Add FFM to distribution module

FDelporte committed:
#614 Removing native libraries (WIP)

FDelporte committed:
#614 Removing native libraries (WIP)

Commit:
Cleanup

Commit:
Cleanup

Commit:
Cleanup

Commit:
Update GitHub workflows

Commit:
Update GitHub workflows: use Zulu 25

Commit:
Update GitHub workflows and pom

FDelporte committed:
#614 Removing native libraries (WIP)

Copilot committed:
Initial plan

Copilot committed:
Add descriptions to @param javadoc tags in BuildInfo record

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>

FDelporte committed:
Merge pull request #618 from Pi4J/copilot/sub-pr-617

Fix missing descriptions in @param javadoc tags for BuildInfo record

FDelporte committed:
Completed JavaDoc

Commit:
Only read build.properties once

Commit:
Only read build.properties once (thread safe)

Commit:
Only read build.properties once (thread safe)

Commit:
Workflow changes requested by @DigitalSmile in https://github.com/Pi4J/pi4j/pull/615#issuecomment-3957268972

Commit:
Workflow changes requested by @DigitalSmile in https://github.com/Pi4J/pi4j/pull/615#issuecomment-3957268972

FDelporte committed:
Merge pull request #617 from Pi4J/expose-buildinfo

Expose buildinfo so it can be used in, for example, api.pi4j.com

FDelporte committed:
Validating wiring

FDelporte committed:
Updated wiring diagram and pictures

FDelporte committed:
Updated README for SmokeTests

FDelporte committed:
Merge branch 'develop' into 614-remove-libraries

FDelporte committed:
Added Nick as author

DigitalSmile committed:
#620 Fix shutdown sequence to close file descriptor as last step.

eitch committed:
Merge pull request #621 from Pi4J/620-ffmdigitalinputeventwatcher-error

#620 Fix shutdown sequence to close file descriptor as last step.

Copilot committed:
Initial plan

Copilot committed:
Fix: more than 4 digital inputs not receiving events in ffm-plugin

Primary fix: Changed EventWatcher from virtual threads to platform daemon
threads to avoid carrier thread pinning when blocking native poll() is called.
Virtual threads pinned to the ForkJoinPool carrier threads (one per CPU core)
prevented 5th+ pins from running on Raspberry Pi 4 (4 cores).

Secondary fix: Corrected LineConfigAttribute debounce mask from `1L << bcm`
to `1L` (index 0 in the offsets array), per kernel GPIO documentation. This
fixes debounce being silently ignored on all BCM pins > 0.

Added regression test: testMoreThanFourInputsReceiveEvents verifies that 5+
digital inputs all receive state change events.

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>
Agent-Logs-Url: https://github.com/Pi4J/pi4j/sessions/97e73dd8-a439-48fa-a3c1-c5758cb50f8c

Copilot committed:
test: strengthen regression test to simulate blocking poll() and prove the carrier-thread pinning bug

The mock poll() now blocks each watcher thread (using a CountDownLatch barrier)
until all numPins=5 watchers are simultaneously blocked, directly mimicking
concurrent blocking native poll() syscalls.

With the old virtual-thread code on a <=4-core machine the ForkJoinPool carrier
pool is exhausted and the 5th watcher never enters poll(), causing
allWatchersBlockingLatch.await() to time out — proving the bug existed.

With the fixed platform-daemon-thread code every watcher has its own OS thread,
all 5 block concurrently, and the test completes successfully.

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>
Agent-Logs-Url: https://github.com/Pi4J/pi4j/sessions/d2e6d6d7-e3eb-4563-aec4-1ca7da3e2a40

FDelporte committed:
Adding hardware test for #622

FDelporte committed:
Adding hardware test for #622

FDelporte committed:
Adding hardware test for #622

FDelporte committed:
Adding hardware test for #622

FDelporte committed:
Extra smoke test to validate the problem reported in #622

FDelporte committed:
Merge branch 'develop' into copilot/fix-ffm-plugin-button-handling

# Conflicts:
#	pi4j-test/src/main/java/com/pi4j/test/Main.java
#	pi4j-test/src/main/java/com/pi4j/test/smoketest/FiveDigitalInputsTestCase.java

FDelporte committed:
Merge pull request #623 from Pi4J/copilot/fix-ffm-plugin-button-handling

Fix ffm-plugin: 5th+ digital inputs receive no state change events

Copilot committed:
Initial plan

Copilot committed:
Fix GPIO listener missing RISING events when timestamp LSB is zero

The EventWatcher in FFMDigitalInput was using buf[i] == 0 to detect
empty event slots in the 16-event read buffer. This only checked the
first byte (LSB on little-endian ARM) of the 8-byte timestamp_ns field.
Any valid GPIO event with a nanosecond timestamp divisible by 256 has
a zero LSB, causing it to be silently skipped.

Fix by checking all 8 bytes of the timestamp_ns field via bitwise OR.
A kernel GPIO event timestamp being exactly 0 (system boot time) is
impossible in practice, so this correctly identifies empty slots.

Add regression test testInputEventProcessingWithZeroTimestampLSB that
verifies a RISING event with timestampNs=256 (LSB=0) is dispatched.

Agent-Logs-Url: https://github.com/Pi4J/pi4j/sessions/7ccb581f-c775-4435-b2e4-103785c7d7a6

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>

FDelporte committed:
Merge pull request #625 from Pi4J/copilot/fix-gpio-listener-event-loss

Fix FFM GPIO EventWatcher silently dropping events when timestamp LSB is zero

taartspi committed:
debounce fails. Testing with all merged code hit this change was wrong.  Bit shift is required : the parm is a bitmap identifying the lines to which the attribute applies, * with each bit number corresponding to the index into &struct gpio_v2_line_request.offsets

taartspi committed:
Merge pull request #626 from Pi4J/debounce_fail_PR623

debounce fails. Testing with all merged code hit this change was wron…

taartspi committed:
Create a shell script to run the smoketest.  pi4j-test/assets/runSmoketest.sh

eitch committed:
Merge pull request #627 from Pi4J/smoketest_as_shell_script

Create a shell script to run the smoketest.  pi4j-test/assets/runSmok…

eitch committed:
[Project] Set new version 4.0.0-SNAPSHOT

eitch committed:
[Project] Set new version 4.1.0-SNAPSHOT

taartspi committed:
Correct wrong pin number associated with GPIO27

FDelporte committed:
Merge pull request #629 from Pi4J/test_README_update

Correct wrong pin number associated with GPIO27

Copilot committed:
Initial plan

Copilot committed:
Fix OutOfMemory in FFM: use per-call Arena.ofConfined() instead of global Arena.ofAuto()

The root cause was that Pi4JNativeContext.ARENA = Arena.ofAuto() is a static field.
Every MemorySegment allocated via context.allocate*() was associated with this
arena, which lives for the JVM lifetime, so segments were never freed.
Memory grew by ~50MB per 250k I2C calls, eventually causing OOM.

The fix: each native method call in IoctlNative, SMBusNative,
FileDescriptorNative, PermissionNative, and PollNative now creates a fresh
Arena.ofConfined() with try-with-resources, scoping all per-call allocations
(errno capture buffers, data buffers) to the duration of the call.

Pi4JNativeContext.ARENA is retained only for long-lived library lookups in
SMBusContext (SymbolLookup.libraryLookup). CAPTURED_STATE_LAYOUT is exposed
as public static so Native classes can allocate it in their per-call arenas.
The allocateCapturedState() instance method is marked @Deprecated.

Agent-Logs-Url: https://github.com/Pi4J/pi4j/sessions/0934a648-dac1-4363-b861-cb93bb2be3dd

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>

Copilot committed:
Add @Deprecated(since, forRemoval) to allocateCapturedState()

Agent-Logs-Url: https://github.com/Pi4J/pi4j/sessions/0934a648-dac1-4363-b861-cb93bb2be3dd

Co-authored-by: FDelporte <1415873+FDelporte@users.noreply.github.com>

eitch committed:
Merge pull request #630 from Pi4J/copilot/fix-outofmemory-issue

Fix FFM plugin OOM: scope native allocations to per-call Arena.ofConfined()

taartspi committed:
update smoketest readme tables to include LED# as defined in EasyEDA design

taartspi committed:
update smoketest readme tables to include LED# as defined in EasyEDA design

taartspi committed:
Merge branch 'smoketest_readme_add_led#' of https://github.com/Pi4J/pi4j into smoketest_readme_add_led#

FDelporte committed:
Merge pull request #632 from Pi4J/smoketest_readme_add_led#

update smoketest readme tables to include LED# as defined in EasyEDA …

FDelporte committed:
Cleanup

eitch committed:
[Project] Add reminder to publish release on Maven Central

stefanhaustein committed:
see what happens if we drop support for inherited properties

stefanhaustein committed:
remove (now) unused import

FDelporte committed:
Bump to 5.0.0-SNAPSHOT

FDelporte committed:
Merge pull request #615 from Pi4J/614-remove-libraries

Remove native libraries and simplify code base

FDelporte committed:
Maven secrets needed for push

FDelporte committed:
Remove distribution upload, no longer available

FDelporte committed:
Make package to trigger update of the downloads page

stefanhaustein committed:
Merge branch 'Pi4J:develop' into develop

stefanhaustein committed:
Remove the context dependency from config classes, bringing them closer to plain DTOs

stefanhaustein committed:
fix some unused imports

stefanhaustein committed:
unused import cleanup

stefanhaustein committed:
Merge pull request #636 from stefanhaustein/develop

Remove the context dependency from config classes

FDelporte committed:
Correct instructions for running smoke test

stefanhaustein committed:
Add context-less builder factory methods for some missed cases

stefanhaustein committed:
cleanup

stefanhaustein committed:
Well... ¯\_(ツ)_/¯

eitch committed:
Merge pull request #638 from stefanhaustein/develop

Add context-less builder config factory methods for some missed cases

eitch committed:
[Build] Change Maven copy-dependencies phase to 'package'

stefanhaustein committed:
Remove deprecated DeferredDelay (replaced with Delay)

eitch committed:
Merge pull request #640 from stefanhaustein/develop

Remove deprecated DeferredDelay (replaced with Delay)

stefanhaustein committed:
Change the PWM duty cycle to a double

stefanhaustein committed:
Convert the PWM duty cylcle to double. Preserve old setters, as Double can't be directly assigned from int constants (oppsed to double)

stefanhaustein committed:
Merge pull request #644 from stefanhaustein/develop

Change the PWM duty cycle type from Integer to double

DigitalSmile committed:
#643 Major refactoring of moc drivers, add gpio mock driver, fix tests

DigitalSmile committed:
Merge remote-tracking branch 'origin/test-framework' into test-framework

# Conflicts:
#	plugins/pi4j-plugin-ffm/pom.xml
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/common/i2c/SMBusNative.java
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/impl/I2CDirect.java
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/impl/I2CSMBus.java
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/pwm/FFMPwmHardware.java
#	plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/spi/FFMSpi.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CDirectTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CFileTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/I2CSMBusTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/PwmTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/integration/SPITest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/mocks/PermissionHelperMock.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit/GPIOTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit/I2CTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit/PWMTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/unit/SPITest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/GPIOPerformanceTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/I2CPerformanceTest.java
#	plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/jmh/SPIPerformanceTest.java
#	plugins/pi4j-plugin-ffm/src/test/native/i2c/i2c-mock.c
#	plugins/pi4j-plugin-ffm/src/test/native/pwm/pwm-mock.c
#	plugins/pi4j-plugin-ffm/src/test/native/spi/spi-mock.c
#	plugins/pi4j-plugin-ffm/src/test/resources/i2c-setup.sh
#	plugins/pi4j-plugin-ffm/src/test/resources/pwm-setup.sh
#	plugins/pi4j-plugin-ffm/src/test/resources/spi-setup.sh
#	plugins/pi4j-plugin/pom.xml
#	pom.xml

DigitalSmile committed:
#643 CI fix

DigitalSmile committed:
#643 CI fix

DigitalSmile committed:
#643 CI fix

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 test CI

Commit:
add github token

eitch commented:
This PR has gotten out of hand... i can't review it anymore...

DigitalSmile commented:
> This PR has gotten out of hand... i can't review it anymore...

Yeah, sorry about that... I will try to make it more readable and put a TLDR with changes to consider when time will come.

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 test CI

DigitalSmile committed:
#643 JMH test extensions

DigitalSmile committed:
#643 Fix CI

DigitalSmile committed:
#643 Fix CI

DigitalSmile committed:
#643 Fix CI

DigitalSmile committed:
#643 Fix CI

DigitalSmile committed:
#643 Fix CI

DigitalSmile committed:
#643 Added JMH tests

DigitalSmile commented:
Closing this draft and reopen for code review,
